share/p2p/shrexnd/server.go

```diff
@@ -0,0 +1,168 @@
+package shrexnd
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/minio/sha256-simd"
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/eds"
+	"github.com/celestiaorg/celestia-node/share/getters"
+	"github.com/celestiaorg/celestia-node/share/ipld"
+	share_p2p_v1 "github.com/celestiaorg/celestia-node/share/p2p/shrexnd/v1/pb"
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+)
+
+const ndProtocolID = "/shrex/nd/0.0.1"
+
+var log = logging.Logger("shrex/nd")
+
+const (
+	writeTimeout = time.Second * 10
+	readTimeout  = time.Second * 10
+	serveTimeout = time.Second * 10
+)
+
+type Server struct {
+	getter       share.Getter
+	store        *eds.Store
+	host         host.Host
+	writeTimeout time.Duration
+	readTimeout  time.Duration
+	serveTimeout time.Duration
+}
+
+func StartServer(host host.Host, store *eds.Store, getter *getters.IPLDGetter) *Server {
+	srv := &Server{
+		getter:       getter,
+		store:        store,
+		host:         host,
+		writeTimeout: writeTimeout,
+		readTimeout:  readTimeout,
+		serveTimeout: serveTimeout,
+	}
+	host.SetStreamHandler(ndProtocolID, srv.handleNamespacedData)
+	return srv
+}
+
+// Stop stops the server
+func (srv *Server) Stop() {
+	srv.host.RemoveStreamHandler(ndProtocolID)
+}
+
+func (srv *Server) handleNamespacedData(stream network.Stream) {
+	err := stream.SetReadDeadline(time.Now().Add(srv.readTimeout))
+	if err != nil {
+		log.Debugf("server: seting read deadline: %s", err)
+	}
+
+	var req share_p2p_v1.GetSharesByNamespaceRequest
+	_, err = serde.Read(stream, &req)
+	if err != nil {
+		log.Errorw("server: reading request", "err", err)
+		stream.Reset() //nolint:errcheck
+		return
+	}
+
+	err = stream.CloseRead()
+	if err != nil {
+		log.Debugf("server: closing read side of the stream: %s", err)
+	}
+
+	err = validateRequest(req)
+	if err != nil {
+		log.Errorw("server: invalid request", "err", err)
+		stream.Reset() //nolint:errcheck
+		return
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), srv.serveTimeout)
+	defer cancel()
+
+	dah, err := srv.store.GetDAH(ctx, req.RootHash)
+	if err != nil {
+		log.Errorw("server: retrieving DAH for datahash", "err", err)
+		srv.respondInternalError(stream)
+		return
+	}
+
+	shares, err := srv.getter.GetSharesByNamespace(ctx, dah, req.NamespaceId)
+	if err != nil {
+		log.Errorw("server: retrieving shares", "err", err)
+		srv.respondInternalError(stream)
+		return
+	}
+
+	resp := namespacedSharesToResponse(shares)
+	srv.respond(stream, resp)
+}
+
+// validateRequest checks correctness of the request
+func validateRequest(req share_p2p_v1.GetSharesByNamespaceRequest) error {
+	if len(req.NamespaceId) != ipld.NamespaceSize {
+		return fmt.Errorf("incorrect namespace id length: %v", len(req.NamespaceId))
+	}
```

Comment:
we could move the helper function from getter/utils to share and then call it here (but just an idea)

Reply:
It is just a single compare. Feels small enough to keep as is.